src/integration-test/java/com/commercetools/sync/integration/commons/utils/ITUtils.java

```diff
@@ -1,24 +1,173 @@
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
 
+/**
+ * This class is meant only meant for internal use of the library's integration tests.
+ */
 public final class ITUtils {
+    public static final String LOCALISED_STRING_CUSTOM_FIELD_NAME = "backgroundColor";
+    public static final String BOOLEAN_CUSTOM_FIELD_NAME = "invisibleInShop";
+
+
+    /**
+     * This method blocks to create an asset custom Type on the CTP project defined by the supplied
+     * {@code ctpClient}, with the supplied data.
+     * @param typeKey   the type key
+     * @param locale    the locale to be used for specifying the type name and field definitions names.
+     * @param name      the name of the custom type.
+     * @param ctpClient defines the CTP project to create the type on.
+     */
+    public static Type createAssetsCustomType(
+        @Nonnull final String typeKey,
+        @Nonnull final Locale locale,
+        @Nonnull final String name,
+        @Nonnull final SphereClient ctpClient) {
+
+        return createTypeIfNotAlreadyExisting(
+            typeKey, locale, name, ResourceTypeIdsSetBuilder.of().addAssets(), ctpClient);
+    }
+
+    /**
+     * This method blocks to create a custom Type on the CTP project defined by the supplied
+     * {@code ctpClient}, with the supplied data.
+     *
+     * @param typeKey                   the type key
+     * @param locale                    the locale to be used for specifying the type name and field definitions names.
+     * @param name                      the name of the custom type.
+     * @param resourceTypeIdsSetBuilder builds the resource type ids for the created type.
+     * @param ctpClient                 defines the CTP project to create the type on.
+     */
+    public static Type createTypeIfNotAlreadyExisting(
+        @Nonnull final String typeKey,
+        @Nonnull final Locale locale,
+        @Nonnull final String name,
+        @Nonnull final ResourceTypeIdsSetBuilder resourceTypeIdsSetBuilder,
+        @Nonnull final SphereClient ctpClient) {
+
+        return typeExists(typeKey, ctpClient)
+            .orElseGet(() -> {
+                final TypeDraft typeDraft = TypeDraftBuilder
+                    .of(typeKey, LocalizedString.of(locale, name), resourceTypeIdsSetBuilder)
+                    .fieldDefinitions(createCustomTypeFieldDefinitions(locale))
+                    .build();
+                return ctpClient.execute(TypeCreateCommand.of(typeDraft)).toCompletableFuture().join();
+            });
+    }
+
+    private static Optional<Type> typeExists(@Nonnull final String typeKey, @Nonnull final SphereClient ctpClient) {
+        return ctpClient
+            .execute(TypeQueryBuilder.of().predicates(QueryPredicate.of(format("key=\"%s\"", typeKey))).build())
+            .toCompletableFuture()
```

Comment:
just as an optimization: don't join it here, but rather in `createTypeIfNotAlreadyExisting` compose completion stages

Reply:
bbc85ad